sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -364,83 +366,263 @@ struct codec_dai dai_lists[] = {
+	 *   DevId_5		07:00	class_id
+	 */
+	link_id = (link->adr[0] >> 48) & GENMASK(15, 0);
+	mfg_id = (link->adr[0] >> 24) & GENMASK(15, 0);
+	part_id = (link->adr[0] >> 8) & GENMASK(15, 0);
+	class_id = link->adr[0] & GENMASK(7, 0);
+
+	if (link->num_adr == 1) {
+		codec[0].name = devm_kasprintf(dev, GFP_KERNEL,
+					       "sdw:%x:%x:%x:%x", link_id,
+					       mfg_id, part_id, class_id);
+
+		if (!codec[0].name)
+			return -ENOMEM;
+	} else {
+		for (i = 0; i < link->num_adr; i++) {
+			unique_id = (link->adr[i] >> 40) & GENMASK(3, 0);
+			codec[i].name = devm_kasprintf(dev, GFP_KERNEL,
+						       "sdw:%x:%x:%x:%x:%x",
+						       link_id, mfg_id, part_id,
+						       class_id, unique_id);
+
+			if (!codec[i].name)
+				return -ENOMEM;
+		}
+	}
+
+	for (i = 0; i < ARRAY_SIZE(dai_lists); i++)
+		if (part_id == dai_lists[i].id)
+			break;
+
+	if (i == ARRAY_SIZE(dai_lists))
+		return -EINVAL;
+
+	for (j = 0; j < link->num_adr; j++)
+		codec[j].dai_name = dai_lists[i].name;
+
+	/* playback & capture are supported in different dai */
+	if (part_id == 0x711)
+		dai_num++;
+
+	j = 2;
+	for (i = id; i < dai_num + id; i++) {
+		char *name, *cpu_name;
+
+		if (i == id && part_id != 0x715) {
+			name = devm_kasprintf(dev, GFP_KERNEL,
+					      "SDW%d-Playback", link_id);
+			dai_links[i].dpcm_playback = 1;
+		} else {
+			name = devm_kasprintf(dev, GFP_KERNEL,
+					      "SDW%d-Capture", link_id);
+			dai_links[i].dpcm_capture = 1;
+		}
+
+		if (!name)
+			return -ENOMEM;
+
+		cpu_name = devm_kasprintf(dev, GFP_KERNEL,
+					  "SDW%d Pin%d", link_id, j++);
+		if (!cpu_name)
+			return -ENOMEM;
+
+		init_dai_link(dai_links, i, name, cpus, cpu_name,
+			      codec, link->num_adr);
+	}
+
+	if (part_id == 0x711)
+		dai_links[id].init = headset_init;
+
+	if ((link_id == 0x2 || link->num_adr > 1) && part_id == 0x1308)
+		dai_links[id].init = second_spk_init;
+
+	return dai_num;
+}
+
+static int sof_card_dai_links_create(struct device *dev,
+				     struct snd_soc_acpi_mach *mach,
+				     struct snd_soc_card *card)
+{
+	struct snd_soc_dai_link_component *idisp_components;
+	struct snd_soc_acpi_mach_params *mach_params;
+	int ssp_num = 0, sdw_num, hdmi_num = 0;
+	struct snd_soc_dai_link_component *cpus;
+	const struct snd_soc_acpi_link_adr *link;
+	struct snd_soc_dai_link *links;
+	const char *board;
+	int num_links;
+	int i, id = 0;
+	int val;
 
 #if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
-	{
-		.name = "iDisp1",
-		.id = 5,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp1_pin, idisp1_codec, platform),
-	},
-	{
-		.name = "iDisp2",
-		.id = 6,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp2_pin, idisp2_codec, platform),
-	},
-	{
-		.name = "iDisp3",
-		.id = 7,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp3_pin, idisp3_codec, platform),
-	},
+	board = dmi_get_system_info(DMI_BOARD_NAME);
+	if (strstr(board, "TigerLake"))
```

Comment:
I think it's probably better to add an is_tgl quirk function, this is not board dependent but PCH dependent.

Reply:
good idea, I will refine it